// turn this into python with this command:
//   protoc --python_out=. --mypy_out=. MyFS.proto

syntax = "proto2";

package MyFS;

enum MessageType {
  LOGIN = 1;
  LOGIN_RESP = 2;
  LIST_DIR = 3;
  LIST_DIR_RESP = 4;
  REQUEST_FILE = 5;
  REQUEST_FILE_STATUS = 6;
  FILE_CONTENTS = 7;
}

message Login {
  optional string user = 1;
  optional string password = 2;
}

message LoginResp {
  optional bool success = 1;
  optional string reason = 2;
}

message ListDir {
  optional string dirname = 1;
}

enum DirEntryType {
  TYPE_FILE = 1;
  TYPE_DIR = 2;
}
message DirEntry {
  optional string entry_name = 1;
  optional DirEntryType type = 2;
  // only populated for type=TYPE_FILE
  optional int32 size = 3;
}

message ListDirResp {
  optional bool success = 1;
  // only populated if not success
  optional string reason = 2;
  repeated DirEntry entries = 3;
}

message RequestFile {
  optional string filename = 1;
}

message RequestFileStatus {
  optional bool success = 1;
  // only present if not success
  optional string reason = 2;
  // only present if success
  optional int32 size = 3;
}

message FileContents {
  optional int32 position = 1;
  optional int32 block_number = 2;
  optional bytes data = 3;
}

message Client2Server {
  optional MessageType type = 1;
  optional Login login = 2;
  optional RequestFile requestfile = 3;
  optional ListDir listdir = 4;
}

message Server2Client {
  optional MessageType type = 1;
  optional LoginResp loginresp = 2;
  optional RequestFileStatus requestfilestatus = 3;
  optional FileContents filecontents = 4;
  optional ListDirResp listdirresp = 5;
}
